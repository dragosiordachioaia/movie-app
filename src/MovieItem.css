.movie-item img {
  max-width: 100%;
}
