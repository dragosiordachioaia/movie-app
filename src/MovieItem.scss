@import "./variables";

.movie-item {
  padding: 1em;
  padding-top: 0;
  position: relative;
  width: calc(100% / 4);
  box-sizing: border-box;

  img {
    max-width: 100%;
    box-shadow: 0px 7px 40px rgba(0, 0, 0, 0.3);
  }

  .info {
    height: 9em;
    overflow: hidden;
    text-shadow: 0px 3px 6px rgba(0, 0, 0, 0.3);

    .title {
      margin: 0;
      color: #fff;
      font-weight: 700;
      margin-top: 10px;
      font-size: 1.2em;
    }

    .genres {
      margin: 0;
      color: $color-accent-secondary;
      font-weight: 400;
      font-size: 0.9em;
    }
  }

  .rating {
    position: absolute;
    background-color: $color-accent-primary;
    color: #fff;
    transform: translate(-50%, -50%);
    top: 0;
    left: 50%;
    border-radius: 50px;
    width: 3em;
    text-align: center;
    padding: 4px;
  }
}
