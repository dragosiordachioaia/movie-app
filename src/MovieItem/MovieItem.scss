@import "../variables";

.movie-item {
  padding: 0.25em;
  padding-top: 0;
  position: relative;
  width: calc(100% / 2);
  transition: width 500ms;
  transition-timing-function: ease-in-out;

  @media (min-width: $SM) {
    width: calc(100% / 2);
    padding: 0.5em;
    padding-top: 0;
  }

  @media (min-width: $SM + 70px) {
    width: calc(100% / 3);
  }

  @media (min-width: $MD) {
    width: calc(100% / 4);
    padding: 0.5em;
    padding-top: 0;
  }

  @media (min-width: $MD + 100px) {
    width: calc(100% / 5);
  }

  @media (min-width: $LG) {
    width: calc(100% / 6);
  }

  @media (min-width: $XL) {
    width: calc(100% / 7);
  }

  box-sizing: border-box;
  animation-name: fade-up;
  animation-duration: 0.6s;
  animation-fill-mode: forwards;
  animation-iteration-count: 1;
  opacity: 0;

  @keyframes fade-up {
    from {
      top: 2em;
      opacity: 0;
    }
    to {
      top: 0;
      opacity: 1;
    }
  }

  .poster-image {
    display: none;
  }
  .poster-image-placeholder {
    display: block;
  }

  &.loaded {
    .poster-image {
      display: block;
    }
    .poster-image-placeholder {
      display: none;
    }
  }

  img,
  svg {
    max-width: 100%;
    box-shadow: 0px 7px 40px rgba(0, 0, 0, 0.3);
  }

  .info {
    height: 9em;
    overflow: hidden;
    text-shadow: 0px 3px 6px rgba(0, 0, 0, 0.3);

    .title {
      margin: 0;
      color: #fff;
      font-weight: 700;
      margin-top: 10px;
      font-size: 1.2em;
      margin-bottom: 0.1em;
    }

    .genres {
      margin: 0;
      color: $color-accent-secondary;
      font-weight: 400;
      font-size: 0.9em;
    }
  }

  .rating {
    position: absolute;
    background-color: $color-accent-primary;
    color: #fff;
    transform: translate(-50%, -50%);
    top: 0;
    left: 50%;
    border-radius: 50px;
    width: 3em;
    text-align: center;
    padding: 4px;
    padding-bottom: 3px;
    font-weight: 700;
  }
}
