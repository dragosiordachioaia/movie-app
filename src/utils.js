export function getImagePath(imageID, config) {
  if (!config) {
    return null;
  }

  return `${config.secure_base_url}${config.poster_sizes[4]}${imageID}`;
}

export function roundToStep(amount, step) {
  step = 1 / step;
  let returnValue = Math.round(amount * step) / step;
  if (Math.floor(returnValue) === returnValue) {
    return Math.floor(returnValue);
  }
  return returnValue.toFixed(1);
}
