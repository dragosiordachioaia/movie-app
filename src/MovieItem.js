import React, { Component } from "react";

import "./MovieItem.scss";

import { getImagePath } from "./utils";

export default class MovieItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      imageLoaded: false,
    };
    this.displayGenres = this.displayGenres.bind(this);
  }

  displayGenres() {
    return this.props.genres
      .filter(genreData => {
        return this.props.data.genre_ids.includes(genreData.id);
      })
      .map(genreData => genreData.name)
      .join(", ");
  }

  render() {
    return (
      <div className={`movie-item ${this.state.imageLoaded ? "loaded" : ""}`}>
        <span className="rating">{this.props.data.vote_average}</span>
        <img
          className="poster-image"
          src={getImagePath(this.props.data.poster_path, this.props.config)}
          alt={`poster-${this.props.data.title}`}
          onLoad={() => this.setState({ imageLoaded: true })}
        />

        <svg
          className="poster-image-placeholder"
          viewBox="0 0 500 750"
          version="1.1"
        >
          <g
            id="Page-1"
            stroke="none"
            strokeWidth="1"
            fill="none"
            fillRule="evenodd"
          >
            <g id="Artboard" fill="#353D42" fillRule="nonzero">
              <rect id="Rectangle" x="0" y="0" width="500" height="750" />
            </g>
          </g>
        </svg>
        <div className="info">
          <p className="title">{this.props.data.title}</p>
          <p className="genres">{this.displayGenres()}</p>
        </div>
      </div>
    );
  }
}
