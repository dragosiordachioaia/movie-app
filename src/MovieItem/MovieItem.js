import React, { Component } from "react";

import "./MovieItem.scss";

import { getImagePath } from "utils";

export default class MovieItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      imageLoaded: false,
    };
    this.displayGenres = this.displayGenres.bind(this);
    this.displayPlaceholder = this.displayPlaceholder.bind(this);
  }

  displayGenres() {
    // we only want to render genres that actually exist in the list coming from
    // the API, because we those are the only ones we have names for
    return this.props.genres
      .filter(genreData => {
        return (
          !this.props.data.genre_ids ||
          this.props.data.genre_ids.includes(genreData.id)
        );
      })
      .map(genreData => genreData.name)
      .join(", ");
  }

  /**
    We need a placeholder in order to avoid visual glitches while the images load.
    The effect looks best if the placeholder image has the same aspect ratio as
    the real image. I've opted to use an svg for this, but it could have been
    anything else.
  **/
  displayPlaceholder() {
    if (this.state.imageLoaded) {
      return null;
    }

    return (
      <svg
        className="poster-image-placeholder"
        viewBox="0 0 500 750"
        version="1.1"
      >
        <g
          id="Page-1"
          stroke="none"
          strokeWidth="1"
          fill="none"
          fillRule="evenodd"
        >
          <g id="Artboard" fill="#353D42" fillRule="nonzero">
            <rect id="Rectangle" x="0" y="0" width="500" height="750" />
          </g>
        </g>
      </svg>
    );
  }

  /**
    To display the image & placeholder as cleanly as possible, we need an onload
    handler for each image in order to know when they have loaded, so that we
    can remove the placeholder.
  **/
  render() {
    return (
      <div className={`movie-item ${this.state.imageLoaded ? "loaded" : ""}`}>
        <span className="rating">{this.props.data.vote_average}</span>
        <img
          className="poster-image"
          src={getImagePath(this.props.data.poster_path, this.props.config)}
          alt={`poster-${this.props.data.title}`}
          onLoad={() => this.setState({ imageLoaded: true })}
        />

        {this.displayPlaceholder()}
        <div className="info">
          <p className="title">{this.props.data.title}</p>
          <p className="genres">{this.displayGenres()}</p>
        </div>
      </div>
    );
  }
}
