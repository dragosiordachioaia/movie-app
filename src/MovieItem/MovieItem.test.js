import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import MovieItem from "./MovieItem";

// jest.mock("SearchBar/SearchBar", () => "SearchBar");

import * as utils from "utils";
utils.getImagePath = (imageID, config) => {
  return `https://server.com/${imageID}`;
};

let component;
let instance;

beforeEach(() => {
  component = renderer.create(<MovieItem data={{}} genres={[]} />);
  instance = component.getInstance();
});

it("renders correctly when loaded", () => {
  instance.state = {
    imageLoaded: true,
  };
  instance.props = {
    data: {
      genre_ids: [4],
      title: "Awesome new film",
      vote_average: 10,
      poster_path: "/path_to_image.jpg",
    },

    genres: [
      {
        id: 4,
        name: "Action",
      },
    ],
  };
  const result = instance.render();
  expect(result).toMatchSnapshot();
});

it("renders correctly when not loaded", () => {
  instance.state = {
    imageLoaded: false,
  };
  instance.props = {
    data: {
      genre_ids: [4],
      title: "Awesome new film",
      vote_average: 10,
      poster_path: "/path_to_image.jpg",
    },

    genres: [
      {
        id: 4,
        name: "Action",
      },
    ],
  };
  const result = instance.render();
  expect(result).toMatchSnapshot();
});
