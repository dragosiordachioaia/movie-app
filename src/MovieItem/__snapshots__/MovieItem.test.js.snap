// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when loaded 1`] = `
<div
  className="movie-item loaded"
>
  <span
    className="rating"
  >
    10
  </span>
  <img
    alt="poster-Awesome new film"
    className="poster-image"
    onLoad={[Function]}
    src="https://server.com//path_to_image.jpg"
  />
  <div
    className="info"
  >
    <p
      className="title"
    >
      Awesome new film
    </p>
    <p
      className="genres"
    >
      __genres-here__
    </p>
  </div>
</div>
`;

exports[`renders correctly when not loaded 1`] = `
<div
  className="movie-item "
>
  <span
    className="rating"
  >
    10
  </span>
  <img
    alt="poster-Awesome new film"
    className="poster-image"
    onLoad={[Function]}
    src="https://server.com//path_to_image.jpg"
  />
  <svg
    className="poster-image-placeholder"
    version="1.1"
    viewBox="0 0 500 750"
  >
    <g
      fill="none"
      fillRule="evenodd"
      id="Page-1"
      stroke="none"
      strokeWidth="1"
    >
      <g
        fill="#353D42"
        fillRule="nonzero"
        id="Artboard"
      >
        <rect
          height="750"
          id="Rectangle"
          width="500"
          x="0"
          y="0"
        />
      </g>
    </g>
  </svg>
  <div
    className="info"
  >
    <p
      className="title"
    >
      Awesome new film
    </p>
    <p
      className="genres"
    >
      __genres-here__
    </p>
  </div>
</div>
`;
